Report more informative CSS errors

jdm opened the pull request:
This requires https://github.com/servo/rust-cssparser/pull/143 for the final commit. There's no better way to split that work up, unfortunately, and it's extremely easy to bitrot. I would appreciate if we could expedite reviewing this work.

This is the work necessary to enable https://bugzilla.mozilla.org/show_bug.cgi?id=1352669. It makes sense to merge it separately because it's so much effort to keep it up to date with the ongoing Stylo work.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16752)
<!-- Reviewable:end -->

jdm committed:
Report CSS parse errors via enum instead of strings.

jdm committed:
Encapsulate CSS error reporter creation for stylo.

jdm committed:
The ParseErrorReporter trait no longer needs to be Send.

jdm committed:
Thread ParseError return values through CSS parsing.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties/shorthand/inherited_svg.mako.rs, components/style/values/specified/calc.rs, components/style/values/specified/color.rs, components/style/properties/longhand/background.mako.rs, components/style/properties/longhand/text.mako.rs and 61 more
 * @kichjang: components/script/dom/element.rs, components/script/dom/node.rs, components/script_layout_interface/reporter.rs
 * @fitzgen: components/script/dom/element.rs, components/script/dom/node.rs, components/script_layout_interface/reporter.rs
 * @emilio: components/style/properties/shorthand/inherited_svg.mako.rs, components/style/values/specified/calc.rs, components/style/values/specified/color.rs, components/style/properties/longhand/background.mako.rs, components/style/properties/longhand/text.mako.rs and 62 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style and script code, but no tests are modified. Please consider adding a test!

nox commented on components/style/error_reporting.rs:
```diff
@@ -124,3 +124,15 @@ impl ParseErrorReporter for NullReporter {
+#[cfg(feature = "gecko")]
+pub fn create_error_reporter() -> RustLogReporter {
+    RustLogReporter
+}
```
Aren't both of them the exact same function?

jdm replied:
Yes. That changes in my patches to Gecko.

Wafflespeanut commented:
cc @servo/style

emilio commented on components/script_layout_interface/reporter.rs:
```diff
@@ -22,20 +22,20 @@ pub struct CSSErrorReporter {
                   line_offset,
                   location.column,
-                  message)
+                  error.to_string())
```
Perhaps we should implement `Display` for `ContextualParseError`? Fine as a followup I guess.

emilio commented on components/selectors/parser.rs:
```diff
@@ -17,6 +18,17 @@ use std::slice;
+pub enum SelectorParseError<T> {
+    PseudoElementInComplexSelector,
+    NoQualifiedNameInAttributeSelector,
+    TooManyCompoundSelectorComponents,
```
Let's rename this to `TooManyCompoundSelectorComponentsInNegation`, or perhaps `TooComplexNegationSelector`?

emilio commented on components/style/error_reporting.rs:
```diff
@@ -6,22 +6,150 @@
+
+impl<'a> ContextualParseError<'a> {
+    /// Turn a parse error into a string representation.
+    pub fn to_string(&self) -> String {
```
Perhaps we should implement `Display` and implement `to_string` as `format!("{}", self)`? Fine also as a followup/easy issue I guess :)

emilio commented on components/style/properties/longhand/text.mako.rs:
```diff
@@ -175,34 +177,40 @@ ${helpers.single_keyword("unicode-bidi",
+            let result: Result<_, ParseError> = input.try(|input| {
+                match input.expect_ident() {
+                    Ok(ident) => {
+                        (match_ignore_ascii_case! { &ident,
```
I think it shouldn't be needed to wrap it in parenthesis anymore?

jdm replied:
Fixing this will be kind of annoying given how often this pattern appears. I would prefer not to, if that's alright.

emilio replied:
Yeah, that's totally fine for me :)

emilio commented on components/style/values/generics/position.rs:
```diff
@@ -104,23 +104,23 @@ impl<L: HasViewportPercentage> HasViewportPercentage for PositionValue<L> {
             if let Ok(l) = input.try(|i| L::parse(context, i)) {
                 if pos.is_some() {
-                    return Err(())
+                    return Err(StyleParseError::UnspecifiedError.into())
```
Is there an intention to convert all the `UnspecifiedError`s into more explicit error kinds as followup work?

jdm replied:
Yes. It's a nice source of easy issues and a lot of effort for this patch in particular.

emilio commented on components/style/values/mod.rs:
```diff
@@ -237,13 +243,16 @@ pub struct CustomIdent(pub Atom);
+            _ => true
         };
+        if !valid {
+            return Err(BasicParseError::UnexpectedToken(Token::Ident(ident)).into());
```
Perhaps just use braces in the upper `match` branch and get rid of `valid`?

emilio replied:
nvm about this one.

emilio commented on components/style/values/specified/align.rs:
```diff
@@ -340,77 +340,78 @@ impl Parse for JustifyItems {
         "stretch" => Ok(ALIGN_STRETCH),
         _ => Err(())
-    }
+    }).map_err(|()| BasicParseError::UnexpectedToken(Token::Ident(ident)).into())
```
I'm assuming this funky way of doing this is because `ident` is borrowed in the `match` expression, is that right? If so, that's kind of unfortunate, I guess... :(

jdm replied:
Yep! Best compromise I could find.

emilio commented on components/style/values/specified/basic_shape.rs:
```diff
@@ -120,11 +123,12 @@ impl InsetRect {
             Ok(ref s) if s.eq_ignore_ascii_case("inset") =>
                 input.parse_nested_block(|i| GenericInsetRect::parse_function_arguments(context, i)),
-            _ => Err(())
+            _ => Err(BasicParseError::ExpectedToken(Token::Function("inset".into())).into())
```
Do you know if this `"inset".into()` will allocate or not? I'm assuming that since it's an static string it shouldn't need to, but I worry that perhaps to coerce it to do that you may need to be explicit about it. Not a big deal in any case I think

jdm replied:
It should be a Cow::Borrowed afaik.

emilio reviewed (commented):
Wow, that's a big diff. Haven't gone through it with detail, would love help from @nox/@simonsapin reviewing.

Some questions below.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16754) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
Reviewed 9 of 9 files at r1, 4 of 4 files at r2, 1 of 1 files at r3, 75 of 75 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[components/style/custom_properties.rs, line 241 at r4](https://reviewable.io:443/reviews/servo/servo/16752#-KkRWV_K5ERWndP1yyvY:-KkRWV_K5ERWndP1yyvZ:bsfkc9a) ([raw file](https://github.com/servo/servo/blob/8d26cb8ebafd19e79cf41bb3388087ffe4e4f4f2/components/style/custom_properties.rs#L241)):*
> ```Rust
>                 return Err(StyleParseError::CloseSquareBracketInDeclarationValueBlock.into()),
>             Token::CloseCurlyBracket =>
>                 return Err(StyleParseError::CloseCurlyBracketInDeclarationValueBlock.into()),
> ```

These are not just (various kinds of) closing brackets, they are unbalanced / improperly nested ones.

---

*[components/style/media_queries.rs, line 227 at r4](https://reviewable.io:443/reviews/servo/servo/16752#-KkRY-GyzL3KtiiFXlWG:-KkRY-GyzL3KtiiFXlWH:b-c4e3ny) ([raw file](https://github.com/servo/servo/blob/8d26cb8ebafd19e79cf41bb3388087ffe4e4f4f2/components/style/media_queries.rs#L227)):*
> ```Rust
>             Ok(ident) => {
>                 let result: Result<_, ParseError> = MediaQueryType::parse(&*ident)
>                     .map_err(|()| BasicParseError::UnexpectedToken(Token::Ident(ident)).into());
> ```

Could you avoid creating tokens like this? I have plans to change the `Token` data structure (to avoid more memory allocations) that involve structs with private fields, so that only the tokenizer will be able to create tokens. Maybe add a `UnexpectedIdent` error variant instead?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/16752)*
<!-- Sent from Reviewable.io -->

jdm commented:
@SimonSapin I have addressed the review comments and made a new commit which applies to the existing code that you reviewed. I'm going to go ahead and rebase all of my changes separately now, and I won't push them until you are done reviewing the existing code.

SimonSapin commented:
r+

---

Reviewed 54 of 54 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[components/style/values/specified/basic_shape.rs, line 131 at r4](https://reviewable.io:443/reviews/servo/servo/16752#-KjQtC6mKAgMQrEv7NEc-r4-131:-Km3MDxFZQRnJlutud4F:b-qvw9aj) ([raw file](https://github.com/servo/servo/blob/8d26cb8ebafd19e79cf41bb3388087ffe4e4f4f2/components/style/values/specified/basic_shape.rs#L131)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary><blockquote>

It should be a Cow::Borrowed afaik.
</blockquote></details>

Indeed, this uses `Cow::Borrowed` and does not allocate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/16752#-:-Km3N6yN1qB_teoNHFOn:b32wu6)*
<!-- Sent from Reviewable.io -->

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 98d0fb2d07f92a6f9e3356656e7a13be2052f59d with merge d6288b9ee71b1a506daa7e0f23416bb03ffb4613...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2999)

jdm commented:
@bors-servo: r=SimonSapin

bors-servo commented:
:pushpin: Commit 27ae1ef has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin 27ae1ef2e7d78f38e744f2e00e177302f2d6aa82 -->

jdm commented:
@bors-servo: p=1

bors-servo commented:
:hourglass: Testing commit 27ae1ef2e7d78f38e744f2e00e177302f2d6aa82 with merge 061cb5f48e5c93a5decf39e530aea4a566e97341...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7432), [arm32](http://build.servo.org/builders/arm32/builds/7646), [arm64](http://build.servo.org/builders/arm64/builds/7686), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7775), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4531), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4554), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7757), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/242), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/241), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4460), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4527), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/237), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/240), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3002)
Approved by: SimonSapin
Pushing 061cb5f48e5c93a5decf39e530aea4a566e97341 to master...

nox commented:
@jdm What was the problem with not sharing the same `ParseError` across all impls again?

nox commented on components/style/properties/longhand/background.mako.rs:
```diff
@@ -128,17 +128,20 @@ ${helpers.predefined_type("background-image", "ImageLayer",
+            let horizontal = try!(horizontal);
+            let vertical = input.try(RepeatKeyword::parse).ok();
+            Ok(SpecifiedValue::Other(horizontal, vertical))
+        })
```
@jdm Why was that change needed?

nox replied:
Oh yeah borrowing issues, unfortunate. :(

jdm commented:
@nox I'm not sure I understand the question. If it is why rust-cssparser's ParseError does not contain every possible error, the answer is because that's really inconvenient API design.
